R/froll.R

```diff
@@ -1,21 +1,107 @@
-froll = function(fun, x, n, fill=NA, algo=c("fast", "exact"), align=c("right", "left", "center"), na.rm=FALSE, hasNA=NA, adaptive=FALSE) {
-  stopifnot(!missing(fun), is.character(fun), length(fun)==1L, !is.na(fun))
-  algo = match.arg(algo)
+partial2adaptive = function(x, n, align, adaptive) {
+  if (align=="center")
+    stopf("'partial' cannot be used together with align='center'")
+  if (is.list(x) && length(unique(vapply(x, length, 0L)))!=1L)
+    stopf("'partial' does not support variable length of columns in 'x'")
+  if (!(is.numeric(n) || (is.list(n) && all(vapply(n, is.numeric, FALSE)))))
+    stopf("n must be integer vector or list of integer vectors")
+  len = if (is.list(x)) length(x[[1L]]) else length(x)
+  verbose = getOption("datatable.verbose")
+  if (!adaptive) {
+    n = as.list(n) ## test 6006.032
+    if (verbose)
+      cat("partial2adaptive: froll partial=TRUE trimming 'n' and redirecting to adaptive=TRUE\n")
+    trimn = function(n, len, align) {
+      n = min(n, len)
+      if (align=="right")
+        c(seq.int(n), rep(n, len-n))
+      else
+        c(rep(n, len-n), rev(seq.int(n)))
+    }
+    sapply(n, len, align, FUN=trimn, simplify=FALSE)
+  } else {
+    if (!is.list(n)) n = list(n)
+    if (length(unique(vapply(n, length, 0L)))!=1L)
+      stopf("adaptive window provided in 'n' must not to have different lengths")
+    if (length(n[[1L]]) != len)
+      stopf("length of vectors in 'x' must match to length of adaptive window in 'n'")
+    if (verbose)
+      cat("partial2adaptive: froll adaptive=TRUE and partial=TRUE trimming 'n'\n")
+    triman = function(n, align) {
+      if (align=="right")
+        pmin(n, seq_along(n))
+      else
+        pmin(n, rev(seq_along(n)))
+    }
+    sapply(n, align, FUN=triman, simplify=FALSE)
+  }
+}
+
+froll = function(fun, x, n, fill=NA, algo, align=c("right","left","center"), na.rm=FALSE, has.nf=NA, adaptive=FALSE, partial=FALSE, FUN, rho, give.names=FALSE) {
   align = match.arg(align)
-  ans = .Call(CfrollfunR, fun, x, n, fill, algo, align, na.rm, hasNA, adaptive)
+  if (isTRUE(give.names))
+    orig = list(n=n, adaptive=adaptive)
+  if (isTRUE(partial)) {
+    n = partial2adaptive(x, n, align, adaptive)
+    adaptive = TRUE
+  } ## support for partial added in #5441
+  leftadaptive = isTRUE(adaptive) && align=="left"
+  if (leftadaptive) {
+    verbose = getOption("datatable.verbose")
+    rev2 = function(x) if (is.list(x)) sapply(x, rev, simplify=FALSE) else rev(x)
+    if (verbose)
+      cat("froll: adaptive=TRUE && align='left' pre-processing for align='right'\n")
+    x = rev2(x)
```

Comment:
Since we know now how costly negation is, what about reverting in place when possible?

```r
x = rnorm(1e8)
system.time(t1 <- rev(x))
#   user  system elapsed 
#  0.481   0.212   0.693 
system.time(t2 <- reverse(x, copy=TRUE))
#   user  system elapsed 
#  0.117   0.216   0.334 
system.time(t3 <- reverse(x, copy=FALSE))
#   user  system elapsed 
#  0.060   0.000   0.061 
identical(t1, t2)
identical(t1, t3)

x = rnorm(1e8)
n = rep(1e2, 1e8)
system.time(a1 <- frollmax(x, n, align="left", adaptive=TRUE))
#    user  system elapsed 
#  14.330   3.147   7.669
system.time(a2 <- reverse(frollmax(reverse(x, copy=TRUE), reverse(n, copy=TRUE), align="right", adaptive=TRUE), copy=FALSE))
#    user  system elapsed 
#   9.092   0.807   2.687 
identical(a1, a2)
```

Reply:
Good spot. I would say align=left was never a priority to avoid overheads, that's why it has been added as such trick rather than own implementation. Align=left usefulness is much less than of align=right.
It could still make sense to speed up of course but it's rather low priority. Speed up is already x50 comparing to next available solution. That change would theoretically make it even up to x150.
I will fill up an issue about that as it is nice low hanging fruit.

Reply:
Let me ask @eddelbuettel, a quant, who may know subject well.

Q: how often do use rolling functions align=left comparing to align=right?

Reply:
BTW. @ben-schwen any idea why rev() is twice slower than reverse(). This sounds like something that could be improved in base R as well. Please link reverse() implementation.

Reply:
> BTW. @ben-schwen any idea why rev() is twice slower than reverse(). This sounds like something that could be improved in base R as well. Please link reverse() implementation.

Looks like `rev.default(x) = function (x) if (length(x)) x[length(x):1L] else x` has to allocate the memory for both (indices + answer) and do the subsetting which costs as much as copying.

Only wrote it for integer and double, but should be easily extendable.
```C
SEXP reverse(SEXP x, SEXP copyArg) {
  int n = LENGTH(x);

  if (!LOGICAL(copyArg)[0]) {
    switch (TYPEOF(x)) {
    case LGLSXP: case INTSXP: {
      int *xd = INTEGER(x);
      for (int i=0; i < n/2; ++i) {
        int temp = xd[i];
        xd[i] = xd[n-1-i];
        xd[n-1-i] = temp;
      }
    } break;
    case REALSXP: {
      double *xd = REAL(x);
      for (int i=0; i < n/2; ++i) {
        double temp = xd[i];
        xd[i] = xd[n-1-i];
        xd[n-1-i] = temp;
      }
    } break;
    default :
        error(_("Type '%s' is not supported"), type2char(TYPEOF(x)));
    }
    return x;
  }

  SEXP ans = PROTECT(allocVector(TYPEOF(x), n));
  switch (TYPEOF(x)) {
  case LGLSXP: case INTSXP: {
    int *xd = INTEGER(x);
    int *ansd = INTEGER(ans);
    for (int i=0; i < n; i++) {
      ansd[i] = xd[n-1-i];
    }

  } break;
  case REALSXP: {
    double *xd = REAL(x);
    double *ansd = REAL(ans);
    for (int i=0; i < n; i++) {
      ansd[i] = xd[n-1-i];
    }
  } break;
  default:
    error(_("Type '%s' is not supported"), type2char(TYPEOF(x)));
  }

  UNPROTECT(1);
  return ans;
}

```

Reply:
Hm, then R team will probably reject new C code to maintain advocating to make length(x):1L an altrep.

We could possibly even consider frev() to be an exported function in DT. I bet there are few places where internally we could use it instead of rev().

Reply:
For future reference, the follow-up issue is #5885